keps/sig-network/2200-externalips-admission/README.md

```diff
@@ -0,0 +1,267 @@
+of users will never use this feature, and making them ALL run a webhook seems
+terrible.
+
+## Motivation
+
+https://github.com/kubernetes/kubernetes/issues/97110
+
+### Goals
+
+Make it possible to disable an insecure feature for the vast majority of users
+very quickly.
+
+### Non-Goals
+
+* Make this the default (breaking change)
+* Make the feature safe to use.
+
+## Proposal
+
+This KEP proposes to add a built-in admission controller
+"DenyServiceExternalIPs", which rejects any CREATE or UPDATE operation which
+adds a new value to `Service.spec.externalIPs`.  Existing values will be
+tolerated.
+
+The number of rejected operations will be exposed by the standard admission
+metrics (`apiserver_admission_controller_admission_duration_seconds_bucket{name="DenyServiceExternalIPs",rejected="true", ...}`).
+
+### User Stories (Optional)
+
+Alice the admin does not want her users using this insecure feature.  She
+enabled this admission controller and knows no user can use it.  She can then
+audit existing users and make them stop.
+
+### Risks and Mitigations
+
+Some installations may want to use this feature in a more controlled way.  They
+can use a custom webhook admission controller or a policy controller to enforce
+their own rules.
+
+This is a precedent we should not set lightly.  In this case the VAST majority
+of users do not need this feature and this proposal is very surgical in nature.
+As far as we know, there are few other unprivileged fields with this much
+power anywhere in our API, and most of those already have some form of controls
+on them.
+
+## Design Details
+
+One simple admission controller should be enough to disable this misfeature.
+Unfortunately it can not be on by default (that would be breaking).
+
+This means that platform-providers may need to expose an option to control
+this.  While we generally try to avoid mixing knobs that cluster-users would
+set with knobs that cluster-providers own, it seems reasonable to close this as
+soon as possible and consider better answers when we have more cases to
+generalize from.  See "Alternatives" below for more.
+
+See "Proposal" above.
+
+### Test Plan
+
+* Unit tests to ensure CREATE and UPDATE operations are rejected when adding
+  new `externalIPs`.
+* Unit tests to ensure UPDATE operations allow existing `externalIPs`.
+
+### Graduation Criteria
+
+This feature will debut as "GA", bypassing alpha and beta.  It's already opt-in
+and very small scope.
+
+### Upgrade / Downgrade Strategy
+
+Cluster upgrades/downgrades should not be an issue.
+
+### Version Skew Strategy
+
+N/A
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+* **How can this feature be enabled / disabled in a live cluster?**
+  - [X] Other flag
+    - Flag name: --enable-admission-plugins (existing)
+
+* **Does enabling the feature change any default behavior?**
+  Yes.  The `externalIPs` field will not be allowed to mutate.
```

Comment:
Can we mutate from "set" to "not set"? Not really a PRR question but I think it's something we should allow.

Reply:
Yes.  Mentioned elsewhere, will repeat here.